blocks/http/include/gnuradio-4.0/http/HttpBlock.hpp

```diff
@@ -0,0 +1,295 @@
+The result is provided on a single output port as a map with the following keys:
+- status: The HTTP status code, usually 200 on success
+- raw-data: The data of the response
+- mime-type: The mime-type of the response
+)"">;
+
+template<typename T>
+class HttpBlock : public gr::Block<HttpBlock<T>, BlockingIO<false>, HttpBlockDoc> {
+private:
+    // used for queuing GET responses for the consumer
+    std::queue<pmtv::map_t> _backlog;
+    std::mutex              _backlog_mutex;
+
+    std::shared_ptr<std::thread> _thread;
+    std::atomic_size_t           _pendingRequests = 0;
+    std::atomic_bool             _shutdownThread  = false;
+    std::binary_semaphore        _ready{ 0 };
+
+#ifndef __EMSCRIPTEN__
+    std::unique_ptr<httplib::Client> _client;
+#endif
+
+#ifdef __EMSCRIPTEN__
+    void
+    queueWorkEmscripten(emscripten_fetch_t *fetch) {
+        pmtv::map_t result;
+        result["mime-type"] = "text/plain";
+        result["status"]    = static_cast<int>(fetch->status);
+        result["raw-data"]  = std::string(fetch->data, static_cast<std::size_t>(fetch->numBytes));
+
+        queueWork(result);
+    }
+
+    void
+    onSuccess(emscripten_fetch_t *fetch) {
+        queueWorkEmscripten(fetch);
+        emscripten_fetch_close(fetch);
+    }
+
+    void
+    onError(emscripten_fetch_t *fetch) {
+        // we still want to queue the response, the statusCode will just not be 200
+        queueWorkEmscripten(fetch);
+        emscripten_fetch_close(fetch);
+    }
+
+    void
+    doRequestEmscripten() {
+        emscripten_fetch_attr_t attr;
+        emscripten_fetch_attr_init(&attr);
+        if (type == RequestType::POST) {
+            strcpy(attr.requestMethod, "POST");
+            if (!parameters.empty()) {
+                attr.requestData     = parameters.c_str();
+                attr.requestDataSize = parameters.size();
+            }
+        } else {
+            strcpy(attr.requestMethod, "GET");
+        }
+
+        // this is needed so that we can call into member functions again, when we receive the Fetch callback
+        attr.userData = this;
+
+        attr.attributes = EMSCRIPTEN_FETCH_LOAD_TO_MEMORY;
+        attr.onsuccess  = [](emscripten_fetch_t *fetch) {
+            auto src = static_cast<HttpBlock<T> *>(fetch->userData);
+            src->onSuccess(fetch);
+        };
+        attr.onerror = [](emscripten_fetch_t *fetch) {
+            auto src = static_cast<HttpBlock<T> *>(fetch->userData);
+            src->onError(fetch);
+        };
+        const auto target = url + endpoint;
+        std::ignore       = emscripten_fetch(&attr, target.c_str());
+    }
+
+    void
+    runThreadEmscripten() {
+        if (type == RequestType::SUBSCRIBE) {
+            while (!_shutdownThread) {
+                // long polling, just keep doing requests
+                std::thread thread{ &HttpBlock::doRequestEmscripten, this };
```

Comment:
couldn't this reuse threads, instead of creating a new one each time?

Reply:
That's what I attempted first and unfortunately this is not possible. :( There is some strange interplay with the emscripten runtime doing the request, which will cause an infinite loop right here, unless we do it in a separate thread (that we explicitly `join()` on).